types/abitbol.d.ts

```diff
@@ -1,64 +1,158 @@
+
+// #endregion
+
+// #region Special Properties
+
+type ExtractAttributesPropertiesKeys<P> = {
+    [K in keyof P]: P[K] extends (...args: any[]) => any ? never : K;
+}[keyof P]
+
+type ExtractMethodsPropertiesKeys<P> = {
+    [K in keyof P]: P[K] extends (...args: any[]) => any ? K : never;
+}[keyof P]
+
+type AttributesPropertiesKeys<P> = ExtractAttributesPropertiesKeys<P>;
+
+type MethodsPropertiesKeys<P> = ExtractMethodsPropertiesKeys<P>;
+
+type SuperVariables<P> = {
+    // @ts-expect-error A bit hacky but we really need the Class and not the instance here
+    $class: this,
+    $data: Record<string, any>,
     $map: {
-        attributes: {}
-        methods: {}
-        computedProperties: {}
+        attributes: {
+            [K in AttributesPropertiesKeys<P>]: true;
+        },
+        computedProperties: {
+            [K in keyof GetterAndSetter<P> & string] : {
+                get: GetGetterMethodName<K, P>,
+                set: GetSetterMethodName<K, P>,
+                annotations: Record<string, any>,
+            }
+        },
+        methods: {
+            [K in MethodsPropertiesKeys<P>]: {
+                annotations: Record<string, any>,
+            }
+        },
     }
-    $extend<ClassDescriptor>(properties: ClassDescriptor): AbitbolClass<ClassDescriptor & T>
-}
+};
 
-declare type AbitbolObject<Type> = AbitbolComputedProps<Type> & {
-    $class: AbitbolClass<Type>,
-    $map: {
-        attributes: {},
-        methods: {},
-        computedProperties: {},
-    },
-    $data: any,
-    $super(...args): any,
-    $name: string,
-    $computedPropertyName: string,
+type WithSpecialProperties<E, I extends { prototype: any }, K extends keyof E, PI> = PI
+    & (I["prototype"][K] extends CallableFunction ? { $super: I["prototype"][K] } : NonNullable<unknown>)
+    & (K extends IsGetterOrSetterIdentifier<K> ? { $computedPropertyName: GetterOrSetterIdentifier<K> } : NonNullable<unknown>)
+    & { $name: K }
+
+type InjectSpecialProperties<E, I extends { prototype: any }, PI> = {
+    [K in keyof E]: E[K] extends (...args: infer A) => infer R
+        ? (this: WithSpecialProperties<E, I, K, PI>, ...args: A) => R
+        : E[K]
 }
 
-declare namespace Class {
-    function $class(): void
-    namespace $class {
-        namespace $map {
-            const attributes: {}
-            const methods: {}
-            const computedProperties: {}
-        }
-    }
+// #endregion
+
+type ExtractProperties<E, I extends { prototype: any }, SP> = Omit<
+    Omit<I["prototype"], keyof E | keyof SP> & E & ExtractMixins<E, I>,
+    "__include__" | "__classvars__"
+>
+
+type PublicInstance<P, S> = GetterAndSetter<P> & S & P
+
+type PrivateInstance<E, I extends { prototype: any }, PI> = InjectSpecialProperties<E, I, PI>;
+
+export type ExtendedAbitbolClass<E, I, SP, P, S, PI> = Omit<I, keyof E>
+    & { prototype: P }
+    & S
+    & SP
+    & (new (...args: InitArgType<E, I>) => PI)
+
+export class AbitbolClass {
```

Comment:
To be complete, there are missings static $class and static $map

![image](https://github.com/wanadev/abitbol/assets/119947900/16cbdc8b-9c66-4459-adbe-00b482d6ae5e)

Reply:
https://github.com/wanadev/abitbol/pull/44/commits/fe6f3487a38d1789c5c3fb8edce7325ac936650a